File: 150__Evaluate_Reverse_Polish_Notation.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []
        for token in tokens:
            if token not in "+-*/":
                stack.append(int(token))
            else:
                op2 = stack.pop()
                op1 = stack.pop()
                if token == '+':
                    stack.append(op1 + op2)
                elif token == '-':
                    stack.append(op1 - op2)
                elif token == '*':
                    stack.append(op1 * op2)
                else:
                    tmp = op1 / op2
                    if tmp >= 0:
                        stack.append(op1//op2)
                    else:
                        tmp = abs(op1) // abs(op2)
                        stack.append(-tmp)
        return stack[0]

File: Java/150__Evaluate_Reverse_Polish_Notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        for (String token : tokens) {
            if (Character.isDigit(token.charAt(token.length() - 1))) {
                stack.push(Integer.valueOf(token));
            } else {
                int v1 = stack.pop();
                int v2 = stack.pop();
                stack.push(calculate(v2, v1, token));
            }
        }
        return stack.pop();
    }

    private int calculate(int v1, int v2, String operator) {
        switch (operator) {
            case "+":
                return v1 + v2;
            case "-":
                return v1 - v2;
            case "*":
                return v1 * v2;
            case "/":
                return v1 / v2;
        }
        return 0;
    }
}
